eth_dummy_phy/examples/simple/main/emac2emac.c

```diff
@@ -0,0 +1,297 @@
+    gpio_set_intr_type(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, GPIO_INTR_ANYEDGE);
+    gpio_intr_enable(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO);
+    TaskHandle_t task_handle = xTaskGetHandle(pcTaskGetName(NULL));
+    gpio_isr_handler_add(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, gpio_isr_handler, task_handle);
+    ESP_LOGW(TAG, "waiting for RMII CLK sink device interrupt");
+    ESP_LOGW(TAG, "if RMII CLK sink device is already running, reset it by `EN` button");
+    while (1) {
+        ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+        if (gpio_get_level(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO) == 1) {
+            break;
+        }
+    }
+    ESP_LOGI(TAG, "starting Ethernet initialization");
+#else
+    esp_rom_gpio_pad_select_gpio(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO);
+    gpio_set_direction(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, GPIO_MODE_OUTPUT);
+    gpio_set_pull_mode(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, GPIO_FLOATING);
+    gpio_set_level(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, 0);
+    vTaskDelay(pdMS_TO_TICKS(STARTUP_DELAY_MS));
+    gpio_set_level(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, 1);
+#endif // CONFIG_EXAMPLE_RMII_CLK_SOURCE_DEV
+
+    // --- Initialize Ethernet driver ---
+    // Init common MAC and PHY configs to default
+    eth_mac_config_t mac_config = ETH_MAC_DEFAULT_CONFIG();
+    eth_phy_config_t phy_config = ETH_PHY_DEFAULT_CONFIG();
+
+    // Update PHY config based on board specific configuration
+    phy_config.reset_gpio_num = -1; // no HW reset
+
+    // Init vendor specific MAC config to default
+    eth_esp32_emac_config_t esp32_emac_config = ETH_ESP32_EMAC_DEFAULT_CONFIG();
+    // Update vendor specific MAC config based on board configuration
+    // No SMI, speed/duplex must be statically configured the same in both devices
+    esp32_emac_config.smi_mdc_gpio_num = -1;
+    esp32_emac_config.smi_mdio_gpio_num = -1;
+#if CONFIG_EXAMPLE_RMII_CLK_SOURCE_DEV
+    esp32_emac_config.clock_config.rmii.clock_mode = EMAC_CLK_OUT;
+    esp32_emac_config.clock_config.rmii.clock_gpio = EMAC_CLK_OUT_180_GPIO;
+#else
+    esp32_emac_config.clock_config.rmii.clock_mode = EMAC_CLK_EXT_IN;
+    esp32_emac_config.clock_config.rmii.clock_gpio = EMAC_CLK_IN_GPIO;
+#endif // CONFIG_EXAMPLE_RMII_CLK_SOURCE_DEV
+
+    // Create new ESP32 Ethernet MAC instance
+    esp_eth_mac_t *mac = esp_eth_mac_new_esp32(&esp32_emac_config, &mac_config);
+    // Create dummy PHY instance
+    esp_eth_phy_t *phy = esp_eth_phy_new_dummy(&phy_config);
+
+    // Init Ethernet driver to default and install it
+    esp_eth_handle_t eth_handle = NULL;
+    esp_eth_config_t config = ETH_DEFAULT_CONFIG(mac, phy);
+#if CONFIG_EXAMPLE_RMII_CLK_SINK_DEV
+    // REF RMII CLK sink device performs multiple EMAC init attempts since RMII CLK source device may not be ready yet
+    int i;
+    for (i = 1; i <= 5; i++) {
+        ESP_LOGI(TAG, "Ethernet driver install attempt: %i", i);
+        if (esp_eth_driver_install(&config, &eth_handle) == ESP_OK) {
```

Comment:
Maybe it is worth to consider using line bidirectionally. 
First pulse signifying CLK source being ready, second meaning CLK sink is ready.
I made a quick example of how this can look. Maybe it's even worth considering putting it in the phy dummy driver as part of phy init to make establishing connection as straightforward as possible.
```c
#include <stdio.h>
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "driver/gpio.h"
#include "hal/gpio_hal.h"

#define GPIO_PIN    32
//#define BE_MASTER   1

void app_main(void)
{
    setbuf(stdout, NULL);
    gpio_install_isr_service(0);
    #ifdef BE_MASTER
        gpio_set_direction(GPIO_PIN, GPIO_MODE_OUTPUT);
        vTaskDelay(pdMS_TO_TICKS(500));
        // send first pulse meaning i'm ready
        printf("Sending ready pulse...\n");
        gpio_set_level(GPIO_PIN, 1);
        vTaskDelay(pdMS_TO_TICKS(100));
        gpio_set_level(GPIO_PIN, 0);
        // wait for second pulse
        gpio_set_direction(GPIO_PIN, GPIO_MODE_INPUT);
        printf("Waiting for slave's pulse...");
        while(gpio_get_level(GPIO_PIN) == 0){
            vTaskDelay(pdMS_TO_TICKS(25));
        }
        printf("Slave confirmed being ready.\n");
    #else
        // wait for master's pulse
        gpio_set_direction(GPIO_PIN, GPIO_MODE_INPUT);
        printf("Waiting for masters's pulse...");
        while(gpio_get_level(GPIO_PIN) == 0){
            vTaskDelay(pdMS_TO_TICKS(25));
        }
        printf("Master confirmed being ready.\n");
        // wait some time
        vTaskDelay(pdMS_TO_TICKS(500));
        // send ready pulse
        gpio_set_direction(GPIO_PIN, GPIO_MODE_OUTPUT);
        printf("Sending ready pulse...\n");
        gpio_set_level(GPIO_PIN, 1);
        vTaskDelay(pdMS_TO_TICKS(100));
        gpio_set_level(GPIO_PIN, 0);
    #endif
}
```

Reply:
Or not as part of phy init, but to provide two functions in phy dummy header file `clk_sink_do_handshake` and `clk_source_do_handshake` implementing two code blocks I wrote above and those being first functions called.

Reply:
> Maybe it is worth to consider using the line bidirectionally.

It seems to be unnecessary complicated and frankly speaking, I generally don't like the idea of changing GPIO direction on the fly. It has potential to become headache if timing somehow break. And most importantly, don't forget that the current interrupt line is also used to indicate to the source that sink has been reset and so sink could boot to correct mode again.

> Maybe it's even worth considering putting it in the phy dummy driver as part of phy init to make establishing connection as straightforward as possible.

It couldn't go to `dummy_phy` because this signalling is more or less workaround needed just for ESP32. In addition, the same intended behavior can be achieved by hardware means, see [example README](https://github.com/espressif/esp-eth-drivers/pull/21/files#diff-90e3e6df6c9da2f3da2c39ccc13a105bfda6f5b2067dd0382cf414b1129ff2beR49-R56).